docs/using-react-redux/annotating-connected-components-with-flow.md

```diff
@@ -0,0 +1,255 @@
+---
+
+# Annotating Connected Components with Flow
+
+After Flow 0.85, Flow starts [Asking for Required Annotations](https://medium.com/flow-type/asking-for-required-annotations-64d4f9c1edf8) on implicit calls of higher order components within each file import — export cycle. This facilitates Flow to merge type information from file dependencies _before_ it walks the type structure and conducts type inference.
+
+This helps Flow gain significantly better coverage on higher order components. But it also asks that we explicitly annotate the connected components at file exports. Exporting implicit calls of `connect` will raise error:
+
+    Missing type annotation for OP. OP is a type parameter declared in function type [1] and was implicitly
+    instantiated at call of connect [2].
+
+In general, to make Flow happy with connect after 0.85 is a two-phase fix. First, you need to explicitly annotate each connected components at file exports. This shall clear all the “implicitly instantiated” errors. Then, if your codebase contains mismatched types between component definitions and usages, Flow will report those errors after you fix the implicit instantiation errors.
+
+## Fixing the “implicitly instantiated” errors at calls of `connect`
+
+> **Note:** We need `React.AbstractComponent` from Flow v0.89+
+
+### Annotating at function return
+
+The easiest way to annotate connected components is to annotate at function call return. To do this, we need to know to types of props in our components:
+
+- `OwnProps`: likely contain or equal to what you need as the second parameter to `mapStateToProps`. If there are props that are not used by `mapStateToProps`, i.e., the props that "pass through", include them here in `OwnProps` as well
+- `Props`: `OwnProps` plus the props passed in by `mapStateToProps` and `mapDispatchToProps`
+
+> **Note:** Inexact objects don't spread nor `$Diff` very well. It is strongly recommended that you use exact objects for connected components all the time.
+
+```js
+type OwnProps = {|
+  passthrough: string,
+  forMapStateToProps: string
+|}
+
+type Props = {|
+  ...OwnProps,
+  fromMapStateToProps: string,
+  dispatch1: () => void
+|}
+```
+
+With `OwnProps` and `Props` in figured out, we are now ready to annotate the connected components.
+
+In _component definition_, annotate the props with `Props`. The component will have access to all the injected props from `connect`:
+
+```jsx
+import * as React from 'react'
+
+const MyComponent = (props: Props) => (
+  <div onClick={props.dispatch1}>
+    {props.passthrough}
+    {props.fromMapStateToProps}
+  </div>
+)
+```
+
+When we export, this is also when we normally call `connect`, annotate the exported component with _just_ `OwnProps`:
+
+```jsx
+import * as React from 'react'
+
+// const MyComponent = ...
+
+export default (connect(
+  mapStateToProps,
+  mapDispatchToProps
+)(MyComponent): React.AbstractComponent<OwnProps>)
```

Comment:
Type casting should not be needed. In overall type casting should be avoided IMO. You can probably do something like this:

```
export default connect<Props, OwnProps, _, _, _>(
  mapStateToProps,
  mapDispatchToProps
)(MyComponent)
```

Reply:
> In what kind of cases would you prefer casting over type parameters? If there are valid use cases, there might be room for improvement in the libdefs as well.

In our code base we have a lot of nested higher order components alongside connect. So instead of doing this:

```jsx
import { compose } from "redux";
type AllProps = {/** ... */};
type withAProps = {/** ... */};
// and so forth
compose(
  connect(mapState, mapDispatch), // T_T
  withA<$Diff<$Diff<$Diff<AllProps, withDProps>, withCProps>, withBProps>>,
  withB<$Diff<$Diff<AllProps, withDProps>, withCProps>>,
  withC<$Diff<AllProps, withDProps>>,
  withD<AllProps>
)(ComponentWithEverything);
```

We figured we would do this instead:

```jsx
type OwnProps = {/** props essential to this component */};
type AllProps = { // spreaded props should be exact
  ...OwnProps, ...withAProps, ...withBProps, ...withCProps, ...withDProps,
}
const ComponentWithEverything = (props: AllProps) => {/** ... */};
export default (
  compose(withA, withB, withC, withD)
    (ComponentWithEverything): 
      React.AbstractComponent<OwnProps>
);
```

I previously included the description to this situation, but removed it since `$Diff` is tricky. And I put the casting up front because it requires less set up. I am not 100% sure about the best practices though, any suggestions?

Reply:
@jbrown215 what do you think about above? Casting feels quite nasty to me, but the example above looks almost impossible to type. Redux `compose` is an alias to Flows internal compose

Reply:
What's the downside of casting?

I realize that within each of the hocs, providing explicit type parameters helps the hocs to deduce their own return component types, but we're only feeding them to the next layer. So it seems to me that it's not as necessary since all of those calls stay within one file. And given it's not practically possible as we have some big hundreds of components like that 😅, the workaround we figured out was to annotate at function call return.

Reply:
>  the workaround we figured out was to annotate at function call return.

This is the point. Should workarounds be documented as a recommended way of working? Maybe if it's explicitly explained that it's a workaround.

There is a risk that you introduce some unintended discrepancies. see e.g these examples:
https://flow.org/try/#0BTDeEMC4AIEYCYDMBfGoBGB+GA7ArgLboCmATsgJRpTT5FmUDcAUM2DQimlroSeRRZtgFaAF4AfNFCpowGnX6jJ0AG4B7AJYATQUA

Reply:
I actually am a big fan of casting. It has no runtime effect, so it really is just asking flow to ensure that the value matches the type. An equivalent way of doing this without casting is just:
```
// Before
module.exports = (SomeExpr: SomeType);

// After
const exports: SomeType = SomeExpr;
module.exports = exports;
```

IMO, you should just use whichever syntax is easier to express your intent. Sometimes that's explicit type args, other times that's a cast.

Reply:
@villesau I’m trying to understand this example you linked — is it casting the object twice and therefore producing the discrepancies?

Reply:
It's just an example where you can cast the object to something else, but can't actually cast it back to original type. This is totally by design: Inexact objects allows that. But it also makes it possible to make mistakes like in the example. The point is that you might loose some type information with casting if you are not careful.

Reply:
So.. We have notes about inexact objects being tricky, I guess I shall stress more of that? 

And looks like we're not triggering the tricky serial casting thing since the two casting happen on different objects? The first one on annotating the unconnected component, the second time the exported connected component.

Meanwhile, I feel the benefit of annotating nested hocs + connect this way is larger than the potential caveat we're facing. Because we don't have hold of how the other hocs pass types information around. And so I feel annotating once at export is an expressive clean cut, after all simplicity is also helpful to reduce mistakes. We've been doing this with our code base and Flow is able to cover those connected components correctly.